### ISSUE TITLE
programslots: allow unset end date

### ISSUE THREAD

**Issue by equinox0815** [radio-helsinki-graz/pv#10]
As of now a programslots always needs an end date. The reason for this is that timeslots are generated on save and of course those can't be generated until the end of time. The only workaround is to set en end date that is far enough in the future so nobody realizes that they actually end someday. Until #9 is fixed this future end date can't be too far in the future though because in case a show ends or gets moved to another day and or time the user must manually delete any superfluous timeslots which is quite tedious to do.

I propose the following:

- Programslots allow 'not specified' as end date
- In case there is no end date the system generates i.e. 6 months worth of timeslots
- a new manage.py command should be added to generate more timeslots for programslots with no end date (so it can be called by a cron script or the like to produce new timelsot for shows that continue to be on the air)

**Comment by hermannschwaerzlerUIBK** [comment]
In our current system we allocate timeslots until the end of the current year for open end programslots. This makes things easier and is partly due to organisational reasons:
The "makers" of each programme sign an agreement (Sendevereinbarung) that ends at the end of the year and they have to re-sign this agreement at the end of every year for the next year. 
At the end of each year we run a management command that adds timeslots for all the programmes where the agreement for next year is signed. 
So how about this:

- In case there is no end date the system generates timeslots until Dec. 31st. of the current year.
- Add a possibility to save an "agreement signed for YEAR" for a programme.
- Generate new timeslots when the value of this input is changed.

**Comment by equinox0815** [comment]
@hermannschwaerzlerUIBK this is very similar to what we do at the moment so this mode would be ok for us. Altough to be better customizable i propose to make this configurable

Concerning the "aggrement signed" i think it is problematic to make prolongation of shows depended on a signed agreement. Of course in an ideal world it should be treated that way but at least for us the world is far from being ideal and there are a lot of show hosts who sign their agreements too late...

**Comment by nnrcschmdt** [comment]
@equinox0815 and I discussed this a couple of times before :wink: 

I propose:
* to allow until in ProgramSlot to be blank and null,
* to generate timeslots until the end of next quarter or the end of the year if the until date of the program slot is not set,
* add a renew_until_next_quarter action to the admin interface for ProgramSlot,
* add management commands to do both, 
* to add an agreement_signed_for_year field to the show model.

"A perfect world is left to the reader as an exercise" :smile:

**Comment by equinox0815** [comment]
@nnrcschmdt i agree with the first 3 points but what do you mean by **management commands to do both** i only see one command here  _renew_until_next_quarter_ or do you mean one management command for _renew_until_next_quarter_ and one for _renew_until_end_of_year_?

Also how would the _agreement_signed_for_year_ field work? is it a bool or an integer with the actual year, or even an array of years in which the contract has been signed? How is it set and what will the system do with that information?
Afaik is the workflow like this: 

- Somebody wants to do a new show and goes through the normal process of mandatory workshops and other stuff. At the end of the process the person signs a Sendevereinbarung for the year in which the show starts. 
- The show will be added to PV with a null end. PV generates timeslots i.e. until the end of the year.
- When the time comes, say in December of that year the person needs to sign the Sendevereinbarung for the upcoming year.
- When the Sendevereinbarung has been signed the show can be prolonged until the end of the next year. 

What data will be in the _agreement_signed_for_year_ field during each of the steps above and what happens if:

- the Sendevereinbarung does not get signed at all (i'm afraid this happens more often than we like)
- the Sendevereinbarung is not signed in time for the renewal of the show

Will the system allow prolongation in these cases? If yes what does the field do than? If no how can this be enforced to allow the system to follow reality and not the other way round?

And what if a person already signed a Sendevereinbarung for another show? Is this a once per person thing or once per show? I believe we at Radio Helsinki have a so called **Sendungsverantwortliche Person** and this person needs to sign the Sendevereinbarung and afiak this is once per Person but i need to ask our Programmrat to be sure. The problem is that other radio stations might do this differently...

**Comment by equinox0815** [comment]
Ok the Radio Helsinki process is like this:

- Shows have exactly one responsible person (Sendungsverantwortliche Person). There might be other people involved in the production of the show but there is always just one person who needs to sign the contract (Sendevereinbarung).
- There is exactly one contract per show. If a person is responsible for more than one show she/he must sign multiple contracts.
- Contracts are valid until cancelled (bis auf Widerruf). There is not time limit and no need to resign a contract every year.

So in our case prolongation of shows has nothing to do with a contract running out.
